x-pack/legacy/plugins/maps/public/layers/styles/vector/properties/dynamic_color_property.js

```diff
@@ -123,14 +147,104 @@ export class DynamicColorProperty extends DynamicStyleProperty {
     ];
   }
 
-  _getMBColorStops() {
-    if (this._options.useCustomColorRamp) {
-      return this._options.customColorRamp.reduce((accumulatedStops, nextStop) => {
-        return [...accumulatedStops, nextStop.stop, nextStop.color];
-      }, []);
+  _getColorPaletteStops() {
+    if (this._options.useCustomColorPalette && this._options.customColorPalette) {
+      if (isCategoricalStopsInvalid(this._options.customColorPalette)) {
+        return [];
+      }
+
+      const stops = [];
+      for (let i = 1; i < this._options.customColorPalette.length; i++) {
+        const config = this._options.customColorPalette[i];
+        stops.push({
+          stop: config.stop,
+          color: config.color,
+          isDefault: false,
+        });
+      }
+      stops.push({
+        stop: this._options.customColorPalette[0].stop,
+        color: this._options.customColorPalette[0].color,
+        isDefault: true,
+      });
+      return stops;
+    }
+
+    const fieldMeta = this.getFieldMeta();
+    if (!fieldMeta || !fieldMeta.categories) {
+      return [];
+    }
+
+    const colors = getColorPalette(this._options.color);
+    if (!colors) {
+      return [];
+    }
+    const maxLength = Math.min(colors.length, fieldMeta.categories.length + 1);
+
+    const stops = [];
+    for (let i = 0; i < maxLength; i++) {
+      const isDefault = i === maxLength - 1;
+      stops.push({
+        stop: isDefault ? '__DEFAULT__' : fieldMeta.categories[i].key,
+        isDefault: isDefault,
+        color: colors[i],
+      });
+    }
+    return stops;
+  }
+
+  _getMbDataDrivenCategoricalColor() {
+    if (
+      this._options.useCustomColorPalette &&
+      (!this._options.customColorPalette || !this._options.customColorPalette.length)
+    ) {
+      return null;
+    }
+
+    const paletteStops = this._getColorPaletteStops();
+    if (paletteStops.length < 2) {
+      //occurs when no data
+      return null;
+    }
+
+    const mbStops = [];
+    let defaultColor = null;
+    for (let i = 0; i < paletteStops.length; i++) {
```

Comment:
`_getColorPaletteStops` is an internal method and already adds default stop as the last item in the list. This logic is not needed. Just treat the last item in the list as the default color